Fixes #791: Adds WalletMatrix support via "matrix.json" file.

dcentrica opened the pull request:
Adds an informational meta-file about your wallet's features for auto-import into walletmatrix.app. See #791

My1 commented:
is the supported currencies about the currencies supported by the hw itself (including via thirdparty companion wallets) or just with the first party companion wallet?
in the former case the list would get LONG with the ERC20 list, and in the latter case the list still isnt complete.

x1ddos reviewed (commented):
Maybe I missed some. @benma @thisconnect @jadzeidan please double check.

My1 commented on matrix.json:
```diff
@@ -0,0 +1,77 @@
+        "web": "https://shiftcrypto.ch/",
+        "email": "support@shiftcrypto.ch",
+        "location": "Switzerland",
+        "established": "2018-01-23"
```
Maybe 2018 is from shift crypto security which did the bb01 back then?

dcentrica replied:
Fixed.

I don't recall where this date came from. ATM all wallets (apart from CasaApp and potentially BitBox02) are updated manually by us, and would have obtained data from a variety of sources (websites, app-stores, whois queries etc).

My1 commented on matrix.json:
```diff
@@ -0,0 +1,77 @@
+            }
+        ],
+        "wallet_types": ["on-chain"],
+        "security": ["touch-id","fake-wallet"],
```
> These seem to be missing:
> 
> - pin-on-wallet
> - encrypted-password
> - 2fa
> 
> Remove `touch-id`. While BitBox02 does have touch-sensing controls, it doesn't authenticate. According to [the glossary](https://walletmatrix.app/glossary#featuresecurity) it is supposed to authenticate.

Not sure if encrypted password fits tbh since there's no passwords stored on the bb02

x1ddos replied:
Oh right. I read too fast. Was actually thinking about encrypted seed. :+1:

benma replied:
what's a fake wallet :joy:

dcentrica replied:
A "Fake Wallet" is one that only contains a fraction of a person's holdings in order to "palm-off" a would-be attacker that these are your only funds.

My1 commented on matrix.json:
```diff
@@ -0,0 +1,77 @@
+        "fio": "?",
+        "form_factor": "hardware",
+        "custom_node": "Y",
+        "rbf": "?",
```
Not sure if it that even applies to a hw wallet to be honest because it just signs transactions

benma replied:
it still has to support validating and signing them. but other things like "custom_node" are definitely in software wallet territory.

My1 commented on matrix.json:
```diff
@@ -0,0 +1,77 @@
+        "form_factor": "hardware",
+        "custom_node": "Y",
+        "rbf": "?",
+        "psbt": "?",
```
Correct me if I'm wrong but wasn't there a big effort to support psbt a while ago?

x1ddos replied:
Maybe should instead set this to `P`? I assume P is for "planned".

benma replied:
what does it even mean? you can sign PSBT on the bb02 via HWI or Electrum for example. 

@My1 not sure what you are referring to

My1 commented on matrix.json:
```diff
@@ -0,0 +1,77 @@
+        "sna": "Y",
+        "fee_estimation": "?",
+        "fee_control": "?",
+        "hardware": ["secure-element","backup","u2f"],
```
Touch screen not really.

x1ddos replied:
True. Not sure why I even listed it. Removed. Thanks for keeping me honest.

thisconnect replied:
How about adding "Touch sensors"?
The fact that there is almost no mechanical parts that can break is a big `+` IMO, i.e. no button that can break.

dcentrica replied:
By "Touch Sensors" are you referring to the fact that the device has no interactive mechanical parts _at all_ (buttons, switches, drawers, hatches etc)? If so, then this is certainly a candidate new feature for the v4 schema, which I'd tentatively name "non-mechanical".

My1 replied:
yeah it has 2 sensor bars on the edges that respond to touches, I would say aside from the USB-C plug and the card reader there*s not much mechanical interaction.

benma commented:
I think there is a general issue here in that walletmatrix seems to mix hardware wallet and software wallet features. Related discussion: https://github.com/bitcoin-dot-org/Bitcoin.org/issues/3509

thisconnect commented on matrix.json:
```diff
@@ -0,0 +1,77 @@
+                "name": "hardware",
+                "github": "https://github.com/digitalbitbox/",
+                "download": "https://shiftcrypto.ch/bitbox02/",
+                "language": "C"
```
Hasn't almost everything been ported to "RUST"?

benma replied:
true, should be C/Rust.

x1ddos replied:
![image](https://user-images.githubusercontent.com/25405/134146603-92919322-ed4e-4d21-92ce-01fc398460fc.png)

dcentrica replied:
Fixed.

dcentrica commented on matrix.json:
```diff
@@ -0,0 +1,77 @@
+        "fee_control": "?",
+        "hardware": ["secure-element","backup","u2f"],
+        "lightning": [],
+        "refill": [],
```
Are these features _natively_ available within the H/W device itself, or with the mobile-app? There's an important distinction to make between native H/W features, and those available with an accompanying desktop/smartphone app. 

FYI WalletMatrix primarily concerns itself with native features _at the moment_ but the JSON schema is constantly evolving. I see no reason features of a H/W wallet's default/preferred smartphone/desktop app cannot also be associated with a H/W wallet record in future iterations, but it's not supported in the current v3 schema.

My1 replied:
okay then not, the issue was that it was chaotic what is even meant to be in here

dcentrica replied:
Fixed. (And granted you have a point: There is some work to do to map the names used in the schema with the titles in the glossary, FYI: https://walletmatrix.app/glossary#featurerefill)

dcentrica commented:
@My1 

> is the supported currencies about the currencies supported by the hw itself (including via thirdparty companion wallets) or just with the first party companion wallet?
> in the former case the list would get LONG with the ERC20 list, and in the latter case the list still isnt complete.

Fair question! Currencies should be restricted to those that are out of the box / default. In otherwords, if the device comes with a BTC and BCH "app" (Is that how BitBox02 refers to these?), then these should be the two currencies that will be listed.

dcentrica commented on matrix.json:
```diff
@@ -0,0 +1,77 @@
+    "uuid": "4BA68789-FF92-4649-B7D8-65BCE3966589",
+    "version": "2.0",
+    "organisation": {
+        "name": "Shift CryptoAG",
```
Fixed!

dcentrica commented on matrix.json:
```diff
@@ -0,0 +1,77 @@
+        "established": "2018-01-23"
+    },
+    "wallet": {
+        "name": "Bitbox02",
```
Fixed

dcentrica commented on matrix.json:
```diff
@@ -0,0 +1,77 @@
+                "slug": "shiftcryptoHQ"
+            },
+            {
+                "type": "medium",
```
Removed medium. No there is no "blog" entry. The "socials" array is designed to take a pre-defined list of "available" 3rd party platforms ala: Twitter,Discord,Gitter,Medium,Linkedin,Facebook,Telegram,Reddit,Youtube,Instagram, just so that we can easily abstract the logic to determine the URL pattern from a given slug/handle.

Added a complete list of socials from the website (apart from YouTube - the format seems to have changed since last I looked)

dcentrica replied:
> > I think there is a general issue here in that walletmatrix seems to mix hardware wallet and software wallet features. Related discussion: [bitcoin-dot-org/Bitcoin.org#3509](https://github.com/bitcoin-dot-org/Bitcoin.org/issues/3509)
> 
> This is my point.

Understood. The short and sweet answer is that for _hardware_ wallets specifically; then companion _mobile_ and/or _desktop_ wallets and their features are not counted.

As for "applets" that allow a coin to be managed directly within the H/W device itself, that are manufactured, engineered and supplied by the _same_ manufacturer/developer as the device, then these _should_ be counted.

I own a Ledger Nano S, and it has an "ETH" and "BTC" app. From memory (it was some years ago) the "BTC" app/applet came with it, but I had to install the "ETH" one - I could be wrong, it was a long time ago. Regardless, they were produced by Ledger themselves and as far as WalletMatrix is concerned, can be included.

Note that featured currencies/coins are _utterly incidental_ to WalletMatrix, so BitBox02 can list a subset of them, and that will be fine. If Shift Crypto want to maintain _all_ of them within matrix.json going forward, then we're happy for that to happen too :-)

dcentrica replied:
> Is attestation mentioned somewhere aka the feature that companion wallets can verify the bitbox is real?

Can you describe this feature in greater detail? In terms of _verification_ then at the moment searches for Android wallets are run through the WalletScrutiny (https://walletscrutiny.com) service - but not H/W wallet firmware at present (I note that WS has labelled BitBox02 as "Reproducible" so I have just added this feature to the WM backlog!).

My1 replied:
1) ledger nano S and X come with no applets at all (at least the ones I bought last year)
2) okay when it's about the intrinsic features of the HWW itself obviously some things like coin control etc would need to be marked as "doesnt apply" or something like that, and the coin/token list is gonna get LONG if all gets included, personally I would be in favor of only listing "direct" coins and make a mention that these types of tokens are also supported like "ERC20 tokens").
3) regarding the attestation feature it's not the reprod firmware that is its own thing. Attestation means that the HWW has some special keys/certificates in the secure element built into them at manufacturing, that can be passed a challenge for attestation that is then signed (to prove owning the private key) and the application can put the certificate used against a list of known ones and when they match you can be pretty sure your bitbox is real, if the attestation fails (e.g. because the bitbox is fake, or some weird interaction bugs or by nature of having a developer device) you will get something like this when trying to unlock.
![image](https://user-images.githubusercontent.com/6696524/134462851-0f597417-7684-477f-9221-930fec22fa9b.png)
the ledger in fact has a similar feature when you connect it to ledger live.

dcentrica replied:
1). You're probably right - but TBH the list of coins is incidental (and limited within WalletMatrix - I will publish a list of supported ones (that will probably help figure out which ones you can include anayway!)
2). Let me know if anything needs to be corrected with an inline comment (apart from "coin_control" which I've noted)
3). Wicked, thanks for the detail. I have added it to our draft v4 schema document

dcentrica commented on matrix.json:
```diff
@@ -0,0 +1,77 @@
+                "language": "C"
+            }
+        ]
+    }
```
Fixed.

My1 replied:
considering that it was now cleared up that hww direct features only, coin control can get out again since that's a part of the companion rather than the HWW

dcentrica replied:
Fixed.

dcentrica commented on matrix.json:
```diff
@@ -0,0 +1,77 @@
+        "hd": "Y",
+        "sna": "Y",
+        "fee_estimation": "?",
+        "fee_control": "?",
```
Fixed.

dcentrica commented on matrix.json:
```diff
@@ -0,0 +1,77 @@
+        "reproducible_builds": "Y",
+        "hd": "Y",
+        "sna": "Y",
+        "fee_estimation": "?",
```
Fixed.

dcentrica commented on matrix.json:
```diff
@@ -0,0 +1,77 @@
+        "currencies": [
+            {
+                "type": "crypto",
+                "ticker": "BTC,LTC,ETH,BAT"
```
If you can let me know which coins/tokens are supported out of the box, these will be the supported currencies. We're just as keen as you _not_ to have to support that many coins. WalletMatrix is focused on BTC, so any other coin information can be seen as "incidental" for the time being.

My1 replied:
out of the box can be a fun thing. I just remembered that the bitbox (similar to the trezor line) doesnt even come with a firmware so when you literally take it out of the box it supports literally nothing, and only when you install the firmware for the first time you get your coin stuff (but then you get everything as ledger is iirc the only HWW that's this modular)

dcentrica replied:
Understood - I think this is going quite far down the rabbit hole though, I don't think it unreasonable to assume that someone who purchases such a wallet, is capable of flashing the firmware meaning that the sum total of the device's functionality and features can be considered to also comprise those of the firmware.

My1 replied:
yeah it is going down the rabbithole but on the other hand is it unreasonable for people to add the applets to the ledger as that's basically part of their setup procedure (and likely the reason they can support as much as they do)

dcentrica commented:
Apologies, the original PR was mistakenly using the v2 schema and should have been v3 which has been corrected.

@x1ddos

* "HD" can be found in the schema at: `wallets.properties.features.properties.hd`
* Added missing "security" features
* Appreciate the list of additional security features. We're continually adding to our draft v4 schema, but as you'll appreciate, this space is moving so fast, no standard terms yet exist outside of those proposed in BIPs for example. We do need to leave it a while sometimes to see when other wallets implement a similar sounding feature and then pull a name for it out of you-know-where that covers it. So these additional features may well make it into future iterations.
* Added missing "privacy" features
* Sorted languages as requested, renamed ML to MS
* Added "testnet"
* Updated remaining features with "Y", "N", "P" ("Planned") accordingly.
* Added missing "hardware" features

@benma A "Fake Wallet" is one that only contains a fraction of a person's holdings in order to "palm-off" a would-be attacker than thes are your only funds.

My1 commented:
> @My1 
> 
> > is the supported currencies about the currencies supported by the hw itself (including via thirdparty companion wallets) or just with the first party companion wallet?
> > in the former case the list would get LONG with the ERC20 list, and in the latter case the list still isnt complete.
> 
> Fair question! Currencies should be restricted to those that are out of the box / default. In otherwords, if the device comes with a BTC and BCH "app" (Is that how BitBox02 refers to these?), then these should be the two currencies that will be listed.

The bb02 has fixed functionality the question is now whether the companion wallet (bitboxapp/electrum/mew/etc) matters and just tge featureset of the first party companion is counted or all that is possible with the hardware.

But if applets are relevant then ledger would support no coins out of the box lol

My1 commented:
> I think there is a general issue here in that walletmatrix seems to mix hardware wallet and software wallet features. Related discussion: https://github.com/bitcoin-dot-org/Bitcoin.org/issues/3509
> 

This is my point.

My1 commented:
Is attestation mentioned somewhere aka the feature that companion wallets can verify the bitbox is real?

dcentrica commented on matrix.json:
```diff
@@ -0,0 +1,77 @@
+              "ticker": "USDT"
+            }
+        ],
+        "wallet_types": ["on-chain"],
```
Review the list of responses at the bottom - but TL;DR "HD" is listed as a general feature under `properties.wallets.items.properties.features.hd`

dcentrica commented on matrix.json:
```diff
@@ -0,0 +1,77 @@
+        ],
+        "wallet_types": ["on-chain"],
+        "security": ["touch-id","fake-wallet"],
+        "privacy": ["tor"],
```
Fixed.

dcentrica commented on matrix.json:
```diff
@@ -0,0 +1,77 @@
+        "privacy": ["tor"],
+        "languages": [
+            {
+                "iso6392": "EN,FR,DE,ZH,BG,ML,JA,PT,HI,HE,FA,SL,TR,ES,RU"
```
Fixed.

dcentrica commented on matrix.json:
```diff
@@ -0,0 +1,77 @@
+        "multisig": "Y",
+        "testnet": "N",
+        "batchtx": "?",
+        "fio": "?",
```
Fixed.

dcentrica commented:
@My1 as promised, the list of supported currencies and a rationale is posted under "Which coins are supported in matrix.json?" at: https://walletmatrix.app/vendors.

My1 commented:
cool to know, one ambiguity tho would be tokens like USDT which exist on multiple networks, like Ether and Tron or whatever (granted Tron isnt listed in the first place but not sure if for example Ether Classic which is listed has anything relevant in terms of tokens.

dcentrica commented:
@x1ddos @benma @My1 FYI, apologies for the long hiatus. I have since composed a "schema explorer" page where schema keys can be looked-up against a description of their desired contents: https://walletmatrix.app/schema/explorer-v3/.

Note also, there is an invite-only Schema repo where discussions occur as to future iterations of the schema. We're actively discussing schema v4. Perhaps further suggestions for schema changes can be raised there once you're happy with this PR for v3? Let me know who should be invited.

benma commented:
Closing this PR - please see https://github.com/digitalbitbox/bitbox02-firmware/issues/791#issuecomment-1567521401
